### ISSUE TITLE
How to handle collection?

### ISSUE THREAD

**Issue by cekk** [collective/gatsby-source-plone#20]
In gatsby we are going to map Plone types to new node-types with object infos.

Collections are special contents that are useful only for "collect" other types that satisfy some search parameters and present them as a list.

The goal of this plugin is to staticize a Plone site, so i think that we need to reproduce them on gatsby.
How do we could do this? What infos do we need to store in the node?

I have two ideas:

- Store search parameters (criteria) to perform a graphql query when we render that content
- Handle collections like folders, and set a list of node children

@tisto how collections are handled in plone.restapi?

**Comment by ajayns** [comment]
So I've been exploring collections in a Plone site and also checking it out in plone.restapi.

If we're deciding to handle collections like folders, setting a list of node children; In case of the `@search` endpoint method, this wouldn't be a problem. But in case of the recursive algorithm, it may traverse the already traversed content objects in the collection's children path, leading to duplicates of nodes/errors, so in that case, we'll need a way to end tree traversal at a collection while still saving it's node children.

About the search parameter idea, I'm still getting used to GraphQL and so I'll need to research on dynamically setting queries. But I believe then, the plugin will just have to give the data to the Gatsby project via the collection node, and it's the Gatsby project template for collection that'll need to handle this, data retrieval?

**Comment by datakurre** [comment]
Thoughts about collections:

* we don't really need the filter; we should recognize filter fields on the base of type schemas and drop them to hide the graphql related warning we see about them

* instead of we link all the items of the collection as its children
* collection might be paged (similarly to search) so all pages should be read
* collection items should be read last so that we only link children nodes we have from the initial traverse
* collection items should not be traversed further

**Comment by ajayns** [comment]
About collections being paged, you mean to say it's `items` field maybe in batches? If that's the case, won't that be possible for even folders and such, and we need to consider this case in the general processData function itself right?

**Comment by datakurre** [comment]
@ajayns Good point that also folders could be paged.

Actually, any content type could have items in them and be batches similarly to search.

It might be actually easier to just handle them all similarly, but keep book of “seen paths” and not add item to BFS queue if it is already seen.

Efficient data structure for “seen” is using JS object as “seen[path] = true;” and then test existence of attribute.

**Comment by ajayns** [comment]
In the current approach, I can't see how we'll need `seen paths`. The algo traverses children and each child will only have one parent, making them traversed only once. The only exception is for collections which include children from other folders etc, and so as I suggested, we do not traverse them, but instead, just link them to the collection node.

**Comment by datakurre** [comment]
@ajayns But you don't really know what content type is a collection. Sure there is a built-in content type named "Collection", but it could be renamed and a new content type with similar behavior can be created with just a few clicks.

**Comment by datakurre** [comment]
@ajayns Yes. Also I didn't see that coming. Initially I thought that we should somehow detect the existence of query-field to recognize collections. But the nice thing with seen paths is that it is generic and we don't need to think collections as a special case.

I agree that search-traversal probably already does the right thing with collections..

**Comment by ajayns** [comment]
Where do you think would be a good place to add docs for this? Since we've mentioned the algorithm in Recursive traversal, maybe I could add a section saying improvements made to handle collections and show the improved algo which uses seen[path]  and could also mention batching for children.

**Comment by datakurre** [comment]
Sounds ok for now.
